### PR TITLE
Support new Stackdriver resource model in Heapster

### PR DESCRIPTION
**Pull request by kawych**
For new resource model, Heapster will export mostly the same metrics with new names and monitored resource, with following differences:
- Metrics for pod that are simple aggregations of one metric for lower-level resource are not exported (for node there are no such metrics).
- Metrics for disk are removed. They are going to be replaced by volume metrics for pods (in a follow up PR) and ephemeral-storage metrics when ephemeral-storage is introduced as Kubernetes resource.
- New metrics will be exposed: container uptime, container request for cpu and memory, restart_count (in a follow up PR), node allocatable cpu and memory.

### CODE REVIEW

**Review by fgrzadkowski — COMMENTED** [review]
I didn't manage to review everything so send what I got so far. I'll finish the review in the evening or tomorrow.

High level comment is that it'd be easier to review it if you do refactorings in a separate PR from actually adding new logic.

File: metrics/core/metrics.go [hunk 1]
```diff
@@ -675,6 +685,28 @@ var MetricFilesystemAvailable = Metric{
...
 }
 
+var MetricVolumeTotal = Metric{
+	MetricDescriptor: MetricDescriptor{
+		Name:        "volume/total_bytes",
+		Description: "Total bytes reserved for a volume",
+		Type:        MetricGauge,
+		ValueType:   ValueInt64,
+		Units:       UnitsCount,
+		Labels:      metricLabels,
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/core/metrics.go]: This is a descriptor for metric labels. metricLabels descriptor specifies a single label as defined here: https://github.com/kubernetes/heapster/blob/master/metrics/core/labels.go#L120

**Review by crassirostris — COMMENTED** [review]
+1 to what Filip wrote, I think it makes sense to make the refactoring, making the tralslation logic more declarative and then make 3 changes:

-- add a struct or a map new translation rules
-- add a flag
-- use new translation rules if the flag is specified

File: metrics/core/metrics.go [hunk 2]
```diff
@@ -179,6 +179,16 @@ var MetricUptime = Metric{
 	},
 }
 
+var MetricRestartCount = Metric{
+	MetricDescriptor: MetricDescriptor{
+		Name:        "restart_count",
+		Description: "Number of container restarts",
+		Type:        MetricCumulative,
+		ValueType:   ValueInt64,
+		Units:       UnitsMilliseconds,
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/core/metrics.go]: Fixed

File: metrics/core/metrics.go [hunk 3]
```diff
@@ -675,6 +685,28 @@ var MetricFilesystemAvailable = Metric{
 	},
 }
 
+var MetricVolumeTotal = Metric{
+	MetricDescriptor: MetricDescriptor{
+		Name:        "volume/total_bytes",
+		Description: "Total bytes reserved for a volume",
+		Type:        MetricGauge,
+		ValueType:   ValueInt64,
+		Units:       UnitsCount,
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/core/metrics.go]: Done

File: metrics/core/metrics.go [hunk 4]
```diff
@@ -675,6 +685,28 @@ var MetricFilesystemAvailable = Metric{
...
+	},
+}
+
+var MetricVolumeUsage = Metric{
+	MetricDescriptor: MetricDescriptor{
+		Name:        "volume/used_bytes",
+		Description: "Usage bytes reserved for a volume",
+		Type:        MetricGauge,
+		ValueType:   ValueInt64,
+		Units:       UnitsCount,
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/core/metrics.go]: Done

File: metrics/sinks/stackdriver/metadata.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+	memoryNodeDaemonUsedBytesMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/node_daemon/memory/used_bytes",
+	}
+
+	// Old resource model metrics
+
+	uptimeMD = &metricMetadata{
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/metadata.go]: Done

File: metrics/sinks/stackdriver/metadata.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+	diskBytesTotalMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "container.googleapis.com/container/disk/bytes_total",
+	}
+
+	// Sink performance metrics
+
+	requestsSent = prometheus.NewCounterVec(
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/metadata.go]: Moved back to stackdriver.go

File: metrics/sinks/stackdriver/stackdriver.go [hunk 7]
```diff
@@ -42,6 +43,11 @@ const (
 	httpResponseCodeClientTimeout = -1
 )
 
+var (
+	useOldResourceModel = true
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: This is overriden by a flag if set

File: metrics/sinks/stackdriver/stackdriver.go [hunk 8]
```diff
@@ -59,100 +65,6 @@ type metricMetadata struct {
...
-
-	diskBytesTotalMD = &metricMetadata{
-		MetricKind: "GAUGE",
-		ValueType:  "INT64",
-		Name:       "container.googleapis.com/container/disk/bytes_total",
-	}
-
-	// Sink performance metrics
-
-	requestsSent = prometheus.NewCounterVec(
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go [hunk 9]
```diff
@@ -213,15 +143,30 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
...
+				if len(req.TimeSeries) >= maxTimeseriesPerRequest {
+					requests = append(requests, req)
+					req = getReq()
+				}
 			}
-			if len(req.TimeSeries) >= maxTimeseriesPerRequest {
-				requests = append(requests, req)
-				req = getReq()
+		}
+		if useNewResourceModel {
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go [hunk 10]
```diff
@@ -359,6 +304,21 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
 
 	opts := uri.Query()
 
+	if len(opts["use_old_resources"]) >= 1 {
+		var err error
+		useOldResourceModel, err = strconv.ParseBool(opts["use_old_resources"][0])
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go [hunk 11]
```diff
@@ -163,10 +75,20 @@ func (sink *StackdriverSink) Stop() {
 func (sink *StackdriverSink) processMetrics(metricValues map[string]core.MetricValue,
 	timestamp time.Time, labels map[string]string, createTime time.Time) []*sd_api.TimeSeries {
 	timeseries := make([]*sd_api.TimeSeries, 0)
-	for name, value := range metricValues {
-		ts := sink.TranslateMetric(timestamp, labels, name, value, createTime)
-		if ts != nil {
-			timeseries = append(timeseries, ts)
+	if useOldResourceModel {
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: I prefer not doing the refactoring in this PR. After talking to @fgrzadkowski I think it's not that obvious how the code should look like. I've written some example code refactored to a map, which still doesn't look ideal - feel free to talk to me offline and discuss which form is better. In general, I think this is a separate discussion and doesn't belong in this PR.

File: metrics/core/metrics.go [hunk 12]
```diff
@@ -179,6 +179,16 @@ var MetricUptime = Metric{
 	},
 }
 
+var MetricRestartCount = Metric{
```
Comment (crassirostris) [inline metrics/core/metrics.go]: This doesn't belong to this PR, this is not about supporting the new resource model

File: metrics/sinks/stackdriver/metadata.go [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+
+	memoryNodeDaemonUsedBytesMD = &metricMetadata{
+		MetricKind: "GAUGE",
+		ValueType:  "INT64",
+		Name:       "kubernetes.io/node_daemon/memory/used_bytes",
+	}
+
+	// Old resource model metrics
+
+	uptimeLegacyMD = &metricMetadata{
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/metadata.go]: Adding legacy as a prefix, not in the middle of the name helps to distinguish and discourage better

File: metrics/sinks/stackdriver/stackdriver.go [hunk 14]
```diff
@@ -42,85 +42,9 @@ const (
...
-		Name:       "container.googleapis.com/container/disk/bytes_used",
-	}
-
-	diskBytesTotalMD = &metricMetadata{
-		MetricKind: "GAUGE",
-		ValueType:  "INT64",
-		Name:       "container.googleapis.com/container/disk/bytes_total",
-	}
+	useOldResourceModel = true
+	useNewResourceModel = false
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: Why isn't it a part of the `StackdriverSink` struct?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 15]
```diff
@@ -42,85 +42,9 @@ const (
 	httpResponseCodeClientTimeout = -1
 )
 
-type StackdriverSink struct {
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: Why move?

File: metrics/sinks/stackdriver/stackdriver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (crassirostris) [reply metrics/sinks/stackdriver/stackdriver.go]: Since @fgrzadkowski approves, I'm fine with duplication

However, personally I think it's better not to write code, rather than write it and then rewrite once again

File: metrics/sinks/stackdriver/stackdriver.go [hunk 16]
```diff
@@ -163,10 +104,20 @@ func (sink *StackdriverSink) Stop() {
...
+		for name, value := range metricValues {
+			ts := sink.TranslateMetric(timestamp, labels, name, value, createTime)
+			if ts != nil {
+				timeseries = append(timeseries, ts)
+			}
+		}
+	}
+	if useNewResourceModel {
+		for name, value := range metricValues {
+			ts := sink.TranslateNewMetric(timestamp, labels, name, value, createTime)
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: If you use the `legacy` notation when listing metrics, use the same notation here

Old method should be named `LegacyTransalteMetric`, new one -- juse `TranslateMetric`

File: metrics/sinks/stackdriver/stackdriver.go [hunk 17]
```diff
@@ -163,10 +104,20 @@ func (sink *StackdriverSink) Stop() {
...
 	timestamp time.Time, labels map[string]string, createTime time.Time) []*sd_api.TimeSeries {
 	timeseries := make([]*sd_api.TimeSeries, 0)
-	for name, value := range metricValues {
-		ts := sink.TranslateMetric(timestamp, labels, name, value, createTime)
-		if ts != nil {
-			timeseries = append(timeseries, ts)
+	if useOldResourceModel {
+		for name, value := range metricValues {
+			ts := sink.TranslateMetric(timestamp, labels, name, value, createTime)
+			if ts != nil {
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]:
```go
if ts := sink.TranslateMetric(...); ts != nil {
  ...
}
```

File: metrics/sinks/stackdriver/stackdriver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (fgrzadkowski) [reply metrics/sinks/stackdriver/stackdriver.go]: I'd prefer not to introduce such refactoring. First it's not obvious for me, it'll actually help with readability. Secondly, I'd like this PR to be minimal. We can refactor later.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 18]
```diff
@@ -194,16 +145,19 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
 			continue
 		}
 
+		// Hack used only with legacy resource type "gke_container". It is used to represent three
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: `Hack used only with legacy resource type "gke_container"`

How is it represented in the logic? I don't see any mentions of the old resource type

File: metrics/sinks/stackdriver/stackdriver.go [hunk 19]
```diff
@@ -359,6 +326,15 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
 
 	opts := uri.Query()
 
+	err := parseBoolFlag(opts, "use_old_resources", &useOldResourceModel)
+	if err != nil {
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]:
```go
if err := ...; err != nil {
  ...
}
```

File: metrics/sinks/stackdriver/stackdriver.go [hunk 20]
```diff
@@ -431,26 +406,34 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
...
 
-	newMetricSet.MetricValues["memory/bytes_used"] = bytesUsed
-
-	memoryFaults := metricSet.MetricValues[core.MetricMemoryPageFaults.MetricDescriptor.Name].IntValue
-	majorMemoryFaults := metricSet.MetricValues[core.MetricMemoryMajorPageFaults.MetricDescriptor.Name].IntValue
+func (sink *StackdriverSink) computeDerivedMetrics(metricSet *core.MetricSet) *core.MetricSet {
+	newMetricSet := &core.MetricSet{MetricValues: map[string]core.MetricValue{}}
+	usage, usageOK := metricSet.MetricValues[core.MetricMemoryUsage.MetricDescriptor.Name]
+	workingSet, workingSetOK := metricSet.MetricValues[core.MetricMemoryWorkingSet.MetricDescriptor.Name]
+	memoryFaults, memoryFaultsOK := metricSet.MetricValues[core.MetricMemoryPageFaults.MetricDescriptor.Name]
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: Move closer to the place where it's used

File: metrics/sinks/stackdriver/stackdriver.go [hunk 21]
```diff
@@ -571,6 +695,21 @@ func createTimeSeries(resourceLabels map[string]string, metadata *metricMetadata
 	}
 }
 
+func createK8sTimeSeries(resource string, resourceLabels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: Same as above: make consistent with the metric names. New methods should be just methods and old methods should have a legacy prefix

File: metrics/sinks/stackdriver/stackdriver.go [hunk 22]
```diff
@@ -431,50 +391,39 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
...
-	}
-
-	newMetricSet := &core.MetricSet{
-		MetricValues: map[string]core.MetricValue{},
-	}
-
-	newMetricSet.MetricValues["memory/bytes_used"] = bytesUsed
-
+func (sink *StackdriverSink) computeDerivedMetrics(metricSet *core.MetricSet) *core.MetricSet {
+	memoryUsage := metricSet.MetricValues[core.MetricMemoryUsage.MetricDescriptor.Name].IntValue
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Added checks, PTAL. I think we should just skip the metric and dependent derived metrics.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 23]
```diff
@@ -194,16 +116,24 @@ func (sink *StackdriverSink) ExportData(dataBatch *core.DataBatch) {
...
+		// Hack used only with legacy resource type "gke_container". It is used to represent three
+		// Kubernetes resources: container, pod or node. For pods container name is empty, for nodes it
+		// is set to artificial value "machine". Otherwise it stores actual container name.
 		if metricSet.Labels["type"] == core.MetricSetTypeNode {
 			metricSet.Labels[core.LabelContainerName.Key] = "machine"
 		}
 
-		computedMetrics := sink.preprocessMemoryMetrics(metricSet)
+		derivedMetrics := &core.MetricSet{MetricValues: map[string]core.MetricValue{}}
+		if useOldResourceModel {
```
[the remark this answers is not in the document]
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: This is actually a mistake, removed this if.

File: metrics/core/metrics.go (hunk 12 above)
Comment (kawych) [reply metrics/core/metrics.go]: For new resources we have different set of metric. I don't think this PR can be sensibly split into "using new resource model" and "using new set of metrics".

File: metrics/sinks/stackdriver/metadata.go (hunk 13 above)
Comment (kawych) [reply metrics/sinks/stackdriver/metadata.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 15 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: By mistake, when removing and reintroducing code from below.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 14 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Moved there

File: metrics/sinks/stackdriver/stackdriver.go (hunk 17 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 16 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 18 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Clarified the comment a little bit, PTAL

File: metrics/sinks/stackdriver/stackdriver.go (hunk 19 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 20 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 21 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/core/metrics.go (hunk 12 above)
Comment (crassirostris) [reply metrics/core/metrics.go]: This is not an old metric in a new format, this is a totally new metric that never existed and can be added in a separate PR

File: metrics/sinks/stackdriver/stackdriver.go (hunk 18 above)
Comment (crassirostris) [reply metrics/sinks/stackdriver/stackdriver.go]: There's `sink.useOldResourceModel` in the condition, now I see. Sorry if I just missed it before

File: metrics/core/metrics.go (hunk 12 above)
Comment (kawych) [reply metrics/core/metrics.go]: Done

Comment (kawych) [reply metrics/core/metrics.go]: Done.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 24]
```diff
@@ -431,54 +395,62 @@ func CreateStackdriverSink(uri *url.URL) (core.DataSink, error) {
...
+		}
 	}
-	newMetricSet.MetricValues["memory/minor_page_faults"] = minorMemoryFaults
 
 	return newMetricSet
 }
 
-func (sink *StackdriverSink) TranslateLabeledMetric(timestamp time.Time, labels map[string]string, metric core.LabeledMetric, createTime time.Time) *sd_api.TimeSeries {
-	resourceLabels := sink.getResourceLabels(labels)
+func (sink *StackdriverSink) LegacyTranslateLabeledMetric(timestamp time.Time, labels map[string]string, metric core.LabeledMetric, createTime time.Time) *sd_api.TimeSeries {
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver.go]: Synced offline, turns out there's no need in a method for the new resource model: old disk metrics _will not be exported_. However, it requires a mention in the PR description

File: metrics/sinks/stackdriver/stackdriver_test.go [hunk 25]
```diff
@@ -244,8 +268,52 @@ func TestPreprocessMemoryMetrics(t *testing.T) {
...
+	computedMetrics := sink.computeDerivedMetrics(metricSet)
 
 	as.Equal(int64(96), computedMetrics.MetricValues["memory/bytes_used"].IntValue)
 	as.Equal(int64(13), computedMetrics.MetricValues["memory/minor_page_faults"].IntValue)
 }
+
+func TestTranslateMetricSet(t *testing.T) {
+	as := assert.New(t)
+
+	containerUptime := testTranslateMetric(as, generateIntMetric(1000), "uptime", containerLabels, "kubernetes.io/container/uptime")
```
Comment (crassirostris) [inline metrics/sinks/stackdriver/stackdriver_test.go]: Nit: I think it's easier to read when input and output is near each other

**Review by crassirostris — COMMENTED** [review]
LGTM, couple of nits left

Deferring to @fgrzadkowski / @piosz  to apply an lgtm label

File: metrics/sinks/stackdriver/stackdriver.go (hunk 24 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Specified this more precisely in a PR description

File: metrics/sinks/stackdriver/stackdriver_test.go (hunk 25 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver_test.go]: done

**Comment by fgrzadkowski** [comment]
@piosz Can you please review? As discussed offline, since you're back, I think you're much better person to review it then me :)

File: metrics/sinks/stackdriver/stackdriver.go [hunk 26]
```diff
@@ -487,64 +459,180 @@ func (sink *StackdriverSink) TranslateMetric(timestamp time.Time, labels map[str
...
+			ts := createTimeSeries("k8s_container", containerLabels, memoryContainerUsedBytesMD, point)
+			ts.Metric.Labels = map[string]string{
+				"memory_type": "non-evictable",
+			}
+			return ts
+		case core.MetricMemoryRequest.MetricDescriptor.Name:
+			point := sink.intPoint(timestamp, timestamp, value.IntValue)
+			return createTimeSeries("k8s_container", containerLabels, memoryRequestedBytesMD, point)
+		}
+	case core.MetricSetTypePod:
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Why don't you export volume metrics?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 27]
```diff
@@ -556,7 +644,39 @@ func (sink *StackdriverSink) getResourceLabels(labels map[string]string) map[str
 	}
 }
 
-func createTimeSeries(resourceLabels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
+func (sink *StackdriverSink) getContainerResourceLabels(labels map[string]string) map[string]string {
+	return map[string]string{
+		"project_id":     sink.project,
+		"location":       sink.zone,
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Add todo to revisit how location is set.

Currently it's based on the zone where Heapster is running and it could be something else because we run nodes in multiple zone.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 28]
```diff
@@ -556,7 +644,39 @@ func (sink *StackdriverSink) getResourceLabels(labels map[string]string) map[str
...
+func (sink *StackdriverSink) getNodeResourceLabels(labels map[string]string) map[string]string {
+	return map[string]string{
+		"project_id":   sink.project,
+		"location":     sink.zone,
+		"cluster_name": sink.cluster,
+		"node_name":    labels[core.LabelNodename.Key],
+	}
+}
+
+func legacyCreateTimeSeries(resourceLabels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: why not implement this by calling createTimeSeries an passing "gke_container" as resource?

**Review by piosz — COMMENTED** [review]
Looks  good, just a few comments.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 26 above)
Comment (crassirostris) [reply metrics/sinks/stackdriver/stackdriver.go]: It's in the PR description

Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: This will be exported in follow up PR.

Comment (piosz) [reply metrics/sinks/stackdriver/stackdriver.go]: Ups, sorry for not reading:|

**Comment by piosz** [comment]
LGTM, please clean up the commits and we can merge.
